lib/date_time/date_time_core.c

```diff
@@ -53,26 +59,51 @@ static void date_time_core_notify_event(enum date_time_evt_type time_source)
 	}
 }
 
-static void date_time_core_schedule_update(bool check_pending)
+static int date_time_core_schedule_work(int interval)
 {
-	/* (Re-)schedule time update work
-	 * if periodic updates are requested in the configuration.
+	/* If an update is already scheduled, exit to avoid rescheduling it.
+	 * Otherwise set the update_scheduled flag.
 	 */
-	if (CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS > 0) {
-		/* Don't reschedule time update work in some cases,
-		 * e.g. if time is not obtained and the work is already pending.
-		 */
-		if (check_pending && k_work_delayable_is_pending(&date_time_update_work)) {
-			return;
-		}
 
+	if (atomic_set(&update_scheduled, true)) {
```

Comment:
I find it hard to believe that there would not be a way to check whether a work item is scheduled or not. Looking at the API, couldn't we use `k_work_delayable_remaining_get()` here? It returns the number of ticks until the work will be submitted and it should return zero if the work item has not been scheduled.

Reply:
When testing the PR and various scenarios. The nice thing about the atomic flag is when it's cleared making this work nicely. As it stands, I'd keep this because I haven't found a nicer way of checking if the update is already scheduled with the currently implemented conditions when this check should be made.

Reply:
I guess it might be possible to use `k_work_cancel_delayable()` or `k_work_cancel_delayable_sync()` when re-scheduling is required. But, keeping the atomic is fine for me, although the name no longer exactly matches the functionality, if the flag is cleared when the update still remains scheduled.

Reply:
I think the easiest, cleanest solution here is just to rename `update_scheduled` to `reschedule_blocked`.

But do note: @tokangas, by the time `atomic_clear(&atomic_clear);` gets called, it is guaranteed that `date_time_update_work_fn` will end up eventually rescheduling, since all code paths (to my knowledge) result in that.

So the `update_scheduled` name is only briefly inaccurate. But `reschedule_blocked` will always be accurate, so I'll just switch to that.

Otherwise I think this design should behave as desired.

It would definitely be nicer if we could actually inspect the state of the workqueue, but doing so from within the work handler is undefined behavior, so realistically it is quite difficult to do things that way.

Cleaner logic for this I suspect will require multiple event queues, changing `date_time_core_current_check` to happen at schedule-time, not execution time, or simply letting go of the requirement that `date_time_core_current_check` not disturb a previously scheduled item.